Shiplog implemented

agb32 opened the pull request:
Implemented a shiplog as per proposal #1053 
Lua bindings created, and examples added to a few missions.
Others should feel free to add to other missions once completed.

onpon4 commented on dat/events/neutral/npc.lua:
```diff
@@ -248,8 +248,8 @@ function create()
-   jm_chance_min = 0
-   jm_chance_max = 0.25
+   jm_chance_min = 0.25
+   jm_chance_max = 0.35
```
This seems unrelated to the topic of the pull request, no? I disagree with the change; I intentionally made it so that jump messages disappear both to minimize the damaging "clickfest" effect it can create (confine it to the start of the game) and so that you don't get bothered with unnecessary garbage later on in the game. You've in fact changed it here so that jump messages are more common than they ever used to be even at its lowest IIRC.

onpon4 commented on dat/missions/neutral/cargo.lua:
```diff
@@ -117,11 +119,14 @@ function land()
       tk.msg(cargo_land_title, cargo_land_p2[rnd.rnd(1, #cargo_land_p2)]:format( cargo_land_p1[rnd.rnd(1, #cargo_land_p1)], cargo ))
       player.pay(reward)
+      dest_p, dest_s = planet.cur()
+      misn.appendLog("Cargo delivered to " .. dest_p:name() .. ", payment " .. math.floor(reward) .. " credits")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/cargo.lua:
```diff
@@ -105,6 +104,9 @@ function accept()
    end
+   misn.createLog(buildCargoMissionDescription( nil, amount, cargo, destplanet, destsys ),"Trade",0)
+   origin_p, origin_s = planet.cur()
+   misn.appendLog("Cargo accepted on "..origin_p:name() .. " for " .. destplanet:name())
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/cargo.lua:
```diff
@@ -117,11 +119,14 @@ function land()
 end
 
 function abort ()
+   misn.appendLog("Cargo delivery abandoned!")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/cargo_rush.lua:
```diff
@@ -145,6 +145,9 @@ function accept()
    end
+   misn.createLog( buildCargoMissionDescription(cargosize[tier], amount, cargo, destplanet, destsys ), "Trade",0)
+   origin_p, origin_s = planet.cur()
+   misn.appendLog("Cargo accepted on "..origin_p:name() .. " for " .. destplanet:name())
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/cargo_rush.lua:
```diff
@@ -161,10 +164,12 @@ function land()
       if intime then
       -- Semi-random message.
       tk.msg(cargo_land_title, string.format( cargo_land_p2[rnd.rnd(1, #cargo_land_p2)], cargo_land_p1[rnd.rnd(1, #cargo_land_p1)], _(cargo) ))
+      misn.appendLog("Cargo delivered on time to " .. destplanet:name() .. ", payment " .. math.floor(reward) .. " credits")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/cargo_rush.lua:
```diff
@@ -161,10 +164,12 @@ function land()
       -- Semi-random message for being late.
       tk.msg(cargo_land_title, string.format( cargo_land_p3[rnd.rnd(1, #cargo_land_p3)], cargo_land_p1[rnd.rnd(1, #cargo_land_p1)], _(cargo), reward/2, reward ))
       reward = reward / 2
+      misn.appendLog("Cargo delivered late to " .. destplanet:name() .. ", payment " .. math.floor(reward) .. " credits")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/pirbounty_dead.lua:
```diff
@@ -155,10 +155,14 @@ function accept ()
 
+   misn.createLog(misn_title:format( misn_level[level], missys:name() ), "Pirate hunt",0)
+   misn.appendLog("Pirate mission in " .. missys:name() .. " accepted on " .. last_sys:name() )
+
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/pirbounty_dead.lua:
```diff
@@ -185,6 +189,7 @@ function jumpout ()
    jumps_permitted = jumps_permitted - 1
    last_sys = system.cur()
    if not job_done and last_sys == missys then
+      misn.appendLog("Mission failed, pirate escaped")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/pirbounty_dead.lua:
```diff
@@ -200,8 +205,10 @@ function land ()
    if job_done and planet.cur():faction() == paying_faction then
       local pay_text
       if target_killed then
+         misn.appendLog("Payment of " .. math.floor(credits) .. " received for death of pirate")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/pirbounty_dead.lua:
```diff
@@ -200,8 +205,10 @@ function land ()
+         misn.appendLog("Payment of " .. math.floor(credits) .. " received for death of pirate")
          pay_text = pay_kill_text[ rnd.rnd( 1, #pay_kill_text ) ]
       else
+         misn.appendLog("Payment of " .. math.floor(credits) .. " received for capture of pirate")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/pirbounty_dead.lua:
```diff
@@ -227,12 +234,14 @@ function pilot_board ()
       local t = subdue_text[ rnd.rnd( 1, #subdue_text ) ]:format( name )
       tk.msg( subdue_title, t )
       succeed()
+      misn.appendLog("Pirate ship boarded")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/pirbounty_dead.lua:
```diff
@@ -227,12 +234,14 @@ function pilot_board ()
       target_ship:disable() -- Stop it from coming back
       if death_hook ~= nil then hook.rm( death_hook ) end
    else
+      misn.appendLog("Failed to board pirate ship") 
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/pirbounty_dead.lua:
```diff
@@ -264,6 +273,7 @@ function pilot_death( p, attacker )
    if attacker == player.pilot() then
       succeed()
       target_killed = true
+      misn.appendLog("Successfully killed the pirate")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/pirbounty_dead.lua:
```diff
@@ -284,6 +294,7 @@ function pilot_death( p, attacker )
       if top_hunter == nil or player_hits >= top_hits then
          succeed()
          target_killed = true
+	 misn.appendLog("Pirate killed, and you did most damage!")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/pirbounty_dead.lua:
```diff
@@ -295,6 +306,7 @@ function pilot_death( p, attacker )
          hook.timer( 3000, "timer_hail", top_hunter )
          misn.osdDestroy()
       else
+         misn.appendLog("Pirate killed, but not by you")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/pirbounty_dead.lua:
```diff
@@ -320,7 +332,7 @@ function hunter_hail( arg )
    local text = share_text[ rnd.rnd( 1, #share_text ) ]
    tk.msg( share_title, text:format( name ) )
-
+   misn.appendLog("Payment of " .. math.floor(credits) .. "made for helping to kill the pirate")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/pirbounty_dead.lua:
```diff
@@ -381,6 +393,7 @@ function spawn_pirate( param )
          pir_jump_hook = hook.pilot( target_ship, "jump", "pilot_jump" )
          pir_land_hook = hook.pilot( target_ship, "land", "pilot_jump" )
 
+	 misn.appendLog("Target pirate ship sighted")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on dat/missions/neutral/pirbounty_dead.lua:
```diff
@@ -389,6 +402,7 @@ function spawn_pirate( param )
             can_capture = false
          end
       else
+         misn.appendLog("No sign of pirate in expected location, you were too late")
```
Assuming the log is supposed to be a useful reference for important past events, this is just cluttering it up with unnecessary information.

onpon4 commented on src/land.c:
```diff
@@ -999,7 +999,10 @@ void land_genWindows( int load, int changetab )
          npc_generate(); /* Generate bar npc. */
    }
+   /* add to the ship log */
+   shiplog_append(0,land_planet->name);
```
Why do we need this? This again seems to be useless info that just clutters up the log. Travel seems much too mundane to be recording.

agb32 commented:
I was putting in examples of how to use the log - I expect they'll get taken out - but I didn't want to interfere with the main missions and put stuff there, because I'd be treading on people's toes.
So once people have started integrating it into the main missions, we can remove these ones.  To be honest, I wouldn't know what is to be considered important in the main missions so didn't want to interfere there.  I figured the simple missions was the simplest way for me to implement, test, debug, and show people how to do it...

onpon4 commented:
One other thing I didn't notice: you shouldn't use string concatenation, use %s formatting instead. And make the strings translatable by adding `_()` around the string.

> but I didn't want to interfere with the main missions and put stuff there, because I'd be treading on people's toes.

It wouldn't be treading on people's toes to add useful logs to the many unique missions we have. It would be helpful. If you don't do it then someone else will have to. I don't see why Reynir would need to come back to add logs to the Hot Dogs From Space mission when you could just do that here for example. They're just logs.

> To be honest, I wouldn't know what is to be considered important in the main missions

I assume you've played the game before, right? So you're familiar with how at least some of these missions go? Add logs for what's important in some of those missions.

> I figured the simple missions was the simplest way for me to implement, test, debug, and show people how to do it...

All it does is create unnecessary busywork for someone else if you're just planning for someone else to remove it.

You know that the docs support giving a usage example, right? Put the example under the `@usage` tag in the comments for the Lua function. Take a look at the other Lua functions for an example. Actually that's another shortcoming here that I didn't notice before. The `@usage` tag is how you're supposed to give examples, not throwaway garbage in actual missions that you expect others to clean up.

onpon4 commented:
Hey, just wanted to follow up on this. I appreciate the improvement you've made with the last commit; while I haven't tested it myself, it seems to be a good system. That said, I would still like to see the "examples" and unnecessary travel log removed, and a proper example given for the function under the `@usage` tag. You don't have to implement proper usages if you don't want to or don't feel confident enough to do so, but the usages you did implement absolutely should not be there (as you yourself admitted).

I would also like to see the unrelated change removed, though that's a personal preference.

Obviously I don't have the power to merge pull requests anyway, but those are changes I would personally suggest.

🕵️

onpon4 commented:
Ah, one more thing I have to note: you should add "Fixes #1053" to the description / OP of this pull request, that'll auto-close #1053 with the merging of this PR.

🕷️

agb32 commented:
Fixes #1053 

ok - I've been thinking a bit about this, and have implemented the following.
There is now a way for mission scripts to remove a log after a certain time.
For cargo logs, the log is therefore now removed after time has moved on from completing the delivery, e.g. when you take off.  So - once completed, its still there for you to read, until you take off.
For pirate logs, the same thing, if you succeed.  However, if you fail, the log hangs around for 1 period, and is then removed.  
I have been finding the pirate logs actually useful for telling me whether I have killed the pirate, or whether someone else got there first - so in that case, useful for them to hang around for a bit.

Journey logs are also now limited to the last 6 journeys.  I think this is useful, because if you come back to the game after a few days or weeks, you can then look at the log and remind yourself where you've come from...

I've also changed the jump point thing back to 0 and 0.25 - not sure how that got in here - I must have forgot to reset it after playing around trying to see why I wasn't getting jump messages anymore...!

What this now means is that the logs won't get filled up unnecessarily, but that useful info is still there for a bit.  Also for new starters, nice for them to have something in their log initially...

Hope everyone agrees with this?

agb32 commented:
And have added the @usage tags to give examples.

onpon4 commented:
Ship logs that disappear are useless. This isn't a solution to the problem. Travel shouldn't be logged because it isn't useful to log travel.

Frankly, the ability to delete logs at all is a problem in my book.

> Also for new starters, nice for them to have something in their log initially...

Empty logs aren't a problem. Why are you treating them as one? At the start of the game you shouldn't have any logs. The first logs you get should be from the first unique mission you get, probably the Empire recruitment.

🕵️

agb32 commented:
ok - logs removed from those missions...

agb32 commented:
Still need to remove the travel log though - that requires a bit more thought, and its late now...

onpon4 commented:
I've submitted a PR to you, @agb32. I think I've removed the travel log successfully, but this isn't fully tested. Please check. https://github.com/agb32/naev/pull/1

On a side note, I've noticed that the log system generally only works for missions. This is a problem since events might need to log stuff, too. I'd suggest at least adding support for events logging stuff, or better, relocating the log stuff outside of misn and not tying it to mission types (which isn't what we want to do anyway).

Of particular note, I know that the hint for the Shadow Vigil mission - which is in an event - will need to use the ship log so the player can look back on what system they need to go to. I think that was a motivation for the proposal actually.

🕵️

onpon4 commented:
Thanks for the merge. Of course as I mentioned on the PR, but just mentioning it here again: one change that needs to happen here is all those uses of printf you've added need to be changed to the WARN macro (or the DEBUG macro if that is more appropriate).

🕵️

agb32 commented:
printf statements now removed and replaced with WARN

onpon4 commented:
Great! Thanks.

Any thoughts on how to extend this to work outside of missions?

🕵️

agb32 commented:
The text type can be anything - so I think it would work - nothing is pulled in from the mission its called from.
The idea was that logs can be categorized by campaign.  e.g. the shadow runs might have a type "Shadow", and then a different name for each of them - the player would then be able to select by type (showing all of the shadow mission logs), and then select one of these to get the log entries just by name.  So unless I misunderstand, it already has that...

agb32 commented:
I've started on the shiplog in shadowvigil as suggested.  However, I'm sure there is more to be done here, but I'm probably not the person to do it - I don't feel I know the mission well enough to understand the subtleties.

agb32 commented:
But its a start, and once pulled, easier for others to work on.

onpon4 commented:
> The text type can be anything - so I think it would work - nothing is pulled in from the mission its called from.

Ah, cool! That should be good then, it just needs to be moved out of the misn module (since that's only available to missions).

🕷️

onpon4 commented:
Ah, have you given any thought into how to make translation work with this? As it is you don't seem to be doing any gettext translation but logs almost certainly should be translated, just preferably in a way that it won't cause problems like #1077.

🕷️

onpon4 commented:
OK, I'm a bit confused by this "log ID" thing you've got going. So, like, you have a local log ID within missions, and does that determine what gets grouped together or does it use text?

It seems that missions should be ignored altogether and the system should just have a list of string identifiers or something. I'm not sure. Maybe it would be easier to implement without the categories thing and just basically be a list of strings? That would mean you wouldn't have to worry about titles changing and needing to group them together and whatnot. Not sure how to solve the text itself but worst case at least you would have everything consistent and as you saw it the first time.

agb32 commented:
Summary:
A logset is created, with a given type and name.  This will then create a unique log ID for this logset (except for if the overwrite flag is set, in which case, any previous logset with the same type and name will be erased, and the log ID for that reused).

This log ID is then used within a chronological linked list of messages, to identify the log messages that belong to this particular log ID, and hence log name and type.

Within a lua mission, it needs to first create a logset, and then save the log ID for this set.  This is written to the players savefile, as part of the mission, so that when a player reloads, they can still append to this logset.  

Once you have created a logset with a name/type, you only need to know the log ID to be able to append log messages to it - you don't need to know the name/type again (though that can be obtained from the log ID if needed).  And if the title changes in the lua script, it doesn't matter - you just keep the log ID, so the save file is still valid for this mission.  It keeps it consistent.

The categories (type) thing is so that its easy for the player to sort through the logs, and select the one they want.  I think we should keep that.  Would get tiresome otherwise, if searching through a lot of logs.

As for translation - no, I couldn't figure out how to do that.  The log messages are saved as text strings.  The obvious place for translation is within the lua script, while writing the logs.  But then they'd remain in that language, even if the player changed their language.

To move it out of the nlua_misn.c module, I think we would need:
createLog()
appendLog()
and possibly deleteLog() and setRemoveLog().

The createLog would return the logID.  The lua script would then need to save this log ID.
So for example, in a mission script, you'd somehow need to put the log ID into the logid element of the Mission struct defined in mission.h.  Currently, this is done within the misn_createLog function.  But perhaps this should be changed to something like:

In a lua mission script:
logid = shiplog.createLog("log name","type",0)
misn.setShipLogID(logid)                   <- which does mission->logid=logid;
shiplog.appendLog(misn.getShipLogID(), "message" )

Or something like that.  Currently, the getting and setting of the logID is hidden within the nlua_misn.c, so a bit more user friendly.

The alternative is to keep nlua_misn.c as it is, but then add a similar functionality to e.g. nlua_evt.c.  Though I have to admin, I'm not entirely clear what the difference between missions and events it - everything I've looked at (shipping, pirate, shadowrun) seems to be a mission.

Hope thats not too confusing (I've just realised I've written a lot!)

onpon4 commented:
So, those IDs are mission-specific. What about replacing that with a global string-based identifier? Strings that would always be untranslated and wouldn't be displayed to the player.

Events and missions are similar but different. There's also Lua code outside of both. Only missions import the misn module, and only events import the evt module. The log should be somewhere else. Maybe the player module? That could be a good fit.

🕷️

agb32 commented:
No, the IDs aren't mission specific.  They just increment from 0 up to 2 billion (and I figured no one would ever get 2 billion missions!).

I guess what you mean is that a mission log should be identifiable from the name and type - and therefore there would be no need to save a log ID in the save file.  Rather we could have something such as:
misn.appendLog("shadow","shadow vigil","some log message here")
(where shadow is the type, shadowvigil is the mission name).
Then that way, there would be no need to save the logID in the mission file.

onpon4 commented:
If the log ID is global, one major problem with using numbers is it's hard to coordinate. How do you find out that numbers 1-20 are taken and 21 is the first available one? For that reason strings would be better than numbers (easier to make sure those are unique).

🕷️

onpon4 commented:
Just to clarify, I do think that IDs is a good practice to help with translation support. That can at least allow categorization to be consistent.

🕷️

agb32 commented:
To find out the first available one, it just scans through the list, takes the max, and increments,  Means that gaps can appear, but we have 2Bn numbers to play with, so not worried.

Note - different players doing the same mission (e.g. shadow vigil) will likely have different log IDs for that mission.  Or indeed, the same player doing a different save file.

One problem with my above suggestion for the lua script is it won't work if titles are mission situation/circumstance dependent.  For example, if we choose to have a mission log for seek and destroy missions, they might have a type "Seek and destroy", while the name might be "Search for pirate xxx, starting from yyy" with xxx being the pirate name, yyy being the system on which the mission was accepted.  In which case, since there isn't a generic title, and this lua script doesn't generate unique missions, it won't be able to get the logid unless it can reconstruct the title (which may be possible).  But simpler to keep the currently implemented approach, and save the logid - less likely to go wrong I think.

So, what I propose:  keep the nlua_misn.c as it is - missions are now done.  But also add a nlua_shiplog.c which can be called be events and other things, which has:
shiplog.createLog(name,type,overwrite)   <- which returns a logID
shiplog.appendLogByID(logid,message)
shiplog.appendLogByName(name,type,message)

Or something like that.

And this is something that can be done at a later stage, once we have an example that requires it... (if you can point me to one, I can look at implementing it).

onpon4 commented on src/nlua_misn.c:
```diff
@@ -1020,3 +1029,95 @@ static int misn_claim( lua_State *L )
+ * @luafunc deleteLog( )
+ * @usage misn.deleteLog()
+ */
+static int misn_deleteLog( lua_State *L )
```
I think you should remove this Lua function, too. Deletion of logs should always be done by the user if it's done at all. Missions shouldn't be in the habit of making missions disappear automatically.

🕵️

onpon4 commented:
> So, what I propose: keep the nlua_misn.c as it is - missions are now done. But also add a nlua_shiplog.c which can be called be events and other things

Why have two separate APIs for the same thing? Just keep it simple and move the log stuff out of misn. It doesn't have to ever have any connection to missions.

agb32 commented:
ok - I've mostly done the above.  Have created nlua_shiplog.c/h, which now contains the createLog and appendLog functions.
I've also updated the shadowvigil to use these.
And removed these functions from the nlua_misn.c.  
deleteLog function removed. 

I've not changed the messages created in shadowvigil, because I don't think I'd be able to do those very well - better to leave that to someone who knows what they're doing, once this is accepted (likewise for all the other missions and events - given that I liked having a log for killing pirates and shipping, shows that I'm out of touch!).

onpon4 commented on src/nlua_shiplog.c:
```diff
@@ -0,0 +1,134 @@
+ *
+ *    @luatparam string name Name for this log.
+ *    @luatparam string type Type of log (e.g travel, trade, etc, can be anything.)
+ *    @luatparam number overwrite Whether to overwrite existing mission with this logname and logtype.  Warning, removes previous entries of this logname and type if 1, or logs of this type if 2.
```
The meaning of this is unclear; reading this I have no idea what this argument does. Also, you shouldn't use numbers that could be 1 or 2, it would be better to just have two separate booleans (which would also mean that `nil` would be valid).

🕵️

agb32 replied:
ok, I'll split it into 2 bools.
Basically, if a user starts a mission and then aborts it, the logset will remain (which is what we want to happen).
If they then restart it at a later date, the existing log can then be replaced with the new attempt, if overwrite=1.  Or if overwrite=0, then the old logset will remain, and a new one will be created.

Overwrite=2 means that all logsets of this type will be replaced by the new logset.  This is more of an edge case, so less likely to be used.  In fact, I could probably remove it for now.

onpon4 replied:
> If they then restart it at a later date, the existing log can then be replaced with the new attempt, if overwrite=1.

Wait, what? I don't think we want this. Logs should *never* be automatically deleted.

> Overwrite=2 means that all logsets of this type will be replaced by the new logset.

Same here. I think you should remove these options altogether. Logs should be permanent.

🕵️

agb32 replied:
Except if we ever implement logs for throwaway missions.  Might happen in the future, we just don't know.  I'll sort something out.  But for the non-throw-away missions, we can just set overwrite=0...

Is there a concept of flags in the lua?  i.e. in the same way as in c we would & flags.  If so, I could replace overwrite with flags, and then would allow for future expansion, as @PhoenixRiver mentions below.

onpon4 replied:
No, even if you did implement them for throwaway missions, they should be permanent.

Flags aren't something you should be doing in Lua. It's technically possible, but it would be a mess. You can't think of Lua like C, code is written for Lua completely differently. If you needed multiple conditions in one arg, the rest of Naev handles multiple cases or whatever by strings. But in a case like this, booleans would make the most sense.

Again though, I don't think logs should *ever* be removed. They should be permanent. If you're having to remove logs, that says that the logs are unimportant and shouldn't have been made in the first place.

🕵️

onpon4 commented on src/nlua_shiplog.c:
```diff
@@ -0,0 +1,134 @@
+   overwrite = luaL_checkint(L,3);
+
+   /* Create a new shiplog */
+   logid = shiplog_create( logname, logtype, overwrite );
```
It looks like you're using the display names of the log name and log type for differentiation? This is a bad idea, it would be a repeat of the problem we have with news articles. It would be better to have separate strings - ID strings, an additional two arguments - that identify categorization on a programmatic level. The pretty version, which would be translated, should only be used for display.

🕵️

agb32 replied:
No - the logid is used for differentiation, once a log has been created.
The log type and logname are displayed to the user in the info widget, but thats all - and these should be translated.
No other differentiation is required.

agb32 replied:
The logid is independent of the logname and logtype...

onpon4 replied:
Then how do multiple, separate Lua scripts use the same "log"? I've told you this already, there needs to be a way for the Lua scripter to specify *where* the log goes. A numeric, auto-generated "log ID" doesn't cut it because multiple scripts (whether that's multiple missions or any other situations) cannot easily share these. You need *string* identifiers.

🕵️

agb32 replied:
ok - now I see - I wasn't aware of the need for this before.  I have a solution - will push it this evening, but basically, a new createNamedLog function which will create a logID if needed, or return it if it already exists, and that will solve it.

onpon4 replied:
That would work, and I would suggest just making that the only logging function, so that everything uses that. Having one without the feature doesn't really add anything, it would just create confusion.

agb32 replied:
The reason to keep it would be for missions that can be done more than once - and we don't know whether these will implement logs in the future or not - so for now, I think best to keep it.  However, I'll get rid of it if you disagree.

onpon4 replied:
Why can't the missions just make a string up that they'll use? It could be unique and that would be fine. If you have one function it unifies everything and makes it crystal-clear to scripters where commonalities lie as far as categorization.

agb32 replied:
Right, I'll do that... will push later.

agb32 replied:
ok - now pushed - updated the createLog function to allow separate lua scripts get the same logset.

onpon4 replied:
Awesome, thanks!

🕷️

onpon4 commented:
> I would suggest that the number makes more sense in this context and is consistent with how this sort of thing is treated in Unix/Linux system utilities and the like

Lua isn't a Unix/Linux system utility. Using a type-checked, numeric value for this sort of thing goes completely against dynamic typing.

> they would indeed help me recover where I am when I have not been able to play in a while

No they wouldn't, these are throwaway missions you were just doing to make some money. You can see how much money you have and what minor missions you can do now. What minor missions you were doing before is totally irrelevant - except for the fact that different *ships* are suited for different missions, which you can find out about easily on the Info screen.

> possibly the failed version could be wiped once a successful run of the mission/event has been completed?

I would much prefer if it doesn't get wiped at all.

🕵️

PhoenixRiver commented:
> Lua isn't a Unix/Linux system utility. Using a type-checked, numeric value
> for this sort of thing goes completely against dynamic typing.

Are you suggesting that the reasoning for it being done with *nix
system utils is falacious? I don't see you arguing with the actual
point of the increasing complexity it involves for dealing with
different modes for the same utility (log system in this case).

> No they wouldn't, these are throwaway missions you were just doing to make
> some money. You can see how much money you have and what minor missions
> you can do now. What minor missions you were doing before is totally irrelevant
> - except for the fact that different *ships* are suited for different
> missions, which you can find out about easily on the Info screen.

I would hope that you are not trying to tell me how to play the game
and certainly not how my memory works. I am an explorer and I rarely
do missions, esepcailly combat heavy ones. I also have associative
memory where things are chained together so if I last did a series of
commodity runs in one direction I was a) travelling through known
space b) with a direction in mind which gives me a hint as to where I
was going. If my commodity logs are empty then I am exploring local
space and if there are combat logs then I know I should be careful
because clearly there are people who attack me around here. In both of
these cases I have learnt the majority of my situation from those
"useless" logs on one screen in one place and am now back to speed
when otherwise it does take me some time wandering the local systems
and trying to find something I recognise as the last thing I was
interested in - actually a lot of the time these days I do not find it
because I have even less time and am less inclined to spend my time
wondering what it was that I forgot rather than playing the game or
just abandoning, what becomes, many things half way through...
missions included.

>> possibly the failed version could be wiped once a successful run of the
>> mission/event has been completed?
>
> I would much prefer if it doesn't get wiped at all.

This is clearly something that needs to be a option for the player
along with the trade, pirate and travel logs (though since they are
seperately categorised they can already be ignored easily by the
player). Would an binary option to stop the logs from autodeleting
anything be possible in the settings menu? Or maybe a seperate system
on the log screen to be able to change the mode choice for each
category being choosable by the player (with a reasonable default
being set of course). The mode system already exists and adding more
modes would not be so difficult since it is numerical - the hardest
bit would be the wording to convey each mode to the player, maybe a
list to choose from:

Never delete logs in this category
Overwrite duplicate events/missions/etc. when they start
Overwrite duplicate events/missions/etc. when they end
Delete after one [long] time unit
Delete after one [short] time unit
Delete after jump
Do not log this category (takes effect after you close the log screen)

I'm not sure about how difficult all these might be to implement and
it might be necessary to add something to note that changes will only
be reflected from that point on - a help page might be necessary but
I'd suggest that the default err on the side of caution to avoid
necesitating most players having to change anything at all.

Hope this helps,

Johann *<:@)

onpon4 commented:
> Are you suggesting that the reasoning for it being done with *nix system utils is falacious?

No, I'm suggesting that you're comparing apples to steak.

> I don't see you arguing with the actual point of the increasing complexity it involves for dealing with different modes for the same utility

No specific examples given, but I don't think those options should be there anyway. As it is, you're just appealing to fear.

> I would hope that you are not trying to tell me how to play the game and certainly not how my memory works.

Not at all, simply that your memory doesn't matter for the cases you're talking about.

> I am an explorer and I rarely do missions, esepcailly combat heavy ones.

So?

Riddle me this: suppose you don't remember a single thing about the game. You're starting on, say, Secondus Station, you see your map, and you have no idea where you got there. Log shows you've not done anything important.

How is this any worse than just starting a new game and having the exact same circumstance, except that you're on Em 5 instead?

No, really, it doesn't matter. Even if you have complete amnesia, you can make your decision regarding what you're going to do next based on what you see. If all you're doing is exploring the game, then fine, have at it. Just look at your map and decide where you're going to explore next. If you go back to a place you've been before, so what as long as you're having fun?

> if there are combat logs then I know I should be careful

The starmap already does that inherently with the presences display.

> actually a lot of the time these days I do not find it because I have even less time and am less inclined to spend my time wondering what it was that I forgot rather than playing the game

That sounds like exactly what you should be doing. What's the problem?

> This is clearly something that needs to be a option for the player

If you want to let the player delete logs, do so by having an interface for them to do it themselves. The option you suggest for "when to autodelete" is far, far more complicated than is necessary. Logs should not be designed to be deleted. They should be designed to be permanent.

PhoenixRiver commented:
*sigh*

> Not at all, simply that your memory doesn't matter for the cases you're
> talking about.

If I was heading for one par tof the universe to do something but have
to stop midway through due to real life then a log of "unimportnat"
things tells me which way I was going... I fail to understand why you
can't see that.

>> I am an explorer and I rarely do missions, esepcailly combat heavy ones.
> So?

So logs adding only missions adds nothing at all to my game and
incereases complexity for no benefit to me. I thought you were against
that.

> Riddle me this: suppose you don't remember a single thing about the game.
> You're starting on, say, Secondus Station, you see your map, and you have no
> idea where you got there. Log shows you've not done anything important.
> How is this any worse than just starting a new game and having the exact
> same circumstance, except that you're on Em 5 instead?

Because I don't have amnesia ever time I play... I have order
associative memory and I don't like spending my time wandering around
the same planets all the time with my time limited play time. Also you
use an extreme example to deny my clear examples (clarified above).

>> if there are combat logs then I know I should be careful
>
> The starmap already does that inherently with the presences display.

The starmap is no replacement for a log of the ships I have recently
fought; now you are comparing apples to oranges as well as suggesting
that graphical and written information are equal... if so why do we go
to the trouble of translating all this written information in game
when we could just draw pictures for it all, aside from people
processing the two types of information differently and with different
competance...

>> actually a lot of the time these days I do not find it because I have even
>> less time and am less inclined to spend my time wondering what it was that
>> I forgot rather than playing the game
>
> That sounds like exactly what you should be doing. What's the problem?

That I want to play the game... implicitly above I failed to be
explicit enough to add and this makes me not want to play the game!


>> This is clearly something that needs to be a option for the player
>
> If you want to let the player delete logs, do so by having an interface for
> them to do it themselves. The option you suggest for "when to autodelete" is
> far, far more complicated than is necessary. Logs should not be designed to
> be deleted. They should be designed to be permanent.

I have an archivist friend who would agree with you and a systems
administrator friend who would disagree with you just as vehemently
(but, to be fair, on his system the logging takes up more drive space
than the user information by a factor of two or three).

You opinion is clearly not shared by everyone but you insist that only
your should be accomodated... I argue for accomodating as many
opinions as possible... I know which would generally be considered the
more reasonable opinion but the decision is the code writer's and the
those with the ability to merge that code.

River

onpon4 commented:
This is giving me a headache. :dizzy_face: Sorry Emmy has been kind of rude. Honestly I just want to get this discussion overwith and have an acceptable solution for everyone. That said, I do think design matters and Emmy is trying to push for good design even if they're not doing so in a particularly tactful way.

So, let's go back to the purpose of the log: the log is supposed to be a reference for important information from the past, right? So, that's mainly useful for recalling past dialog, or in one specific example, it can be used to remind the player of where you need to look to start the Shadow Vigil mission, or of plot details they've learned throughout various campaigns (like information about the Collective).

If I'm understanding, @PhoenixRiver, you're saying that you also find it helpful to have a log of what you were doing last time you were playing, correct? Which isn't in-line with what the ship log system was originally proposed for. That said, there's really no reason I can think of why we can't have both - though I would note that it should be implemented in Lua via an event, probably all contained in a log category called "travel" or something.

As for log deletion, I'm not sure automatic log deletion is particularly helpful. Maybe automatic log pruning for some logs (like limiting the log to X entries) could be useful though? That way you could have a "travel" log (including system entries and combat and such) there, and it would only go back X events, then when new events are added it would get rid of the oldest one (which would prevent it from taking up way more space than it needs to). Emmy has a concern that that would create mistrust or something like that, but we could alleviate that fear by showing in the log when it's a "pruned" log. In practice it would be specific to the travel log.

By the way, just to clarify, I don't want to advocate a return of the C-implemented travel log, rather instead implementing a new travel log via a Lua event (which could also log other interesting things like combat and such).

🕷️

agb32 commented on src/nlua_misn.c:
```diff
@@ -1020,3 +1029,95 @@ static int misn_claim( lua_State *L )
+ *
+ *    @luatparam string Name for this log.
+ *    @luatparam string Type of log (e.g travel, trade, etc, can be anything.)
+ *    @luatparam number Whether to overwrite existing mission with this logname and logtype.  Warning, removes previous entries of this logname and type.
```
Not sure how to do the boolean?  There is no luaL_checkbool or anything like that.
Perhaps I should implement it as a string? e.g. "yes" or "no" etc.

agb32 commented:
I've added a string ID - so missions can now create logs with a name, e.g. "shadow", which identifies this logset.

agb32 commented:
Unclear how lua bools work - there is no luaL_checkbool function or anything like that - what should I be looking for?

onpon4 commented:
`lua_toboolean` is what you want. It's not called "check" because any value is valid as a boolean, just interpreted differently.

🕷️

onpon4 commented:
Oh! By the way, you can mark comments/reviews as resolved once you finish them. Should help with the long length of this PR page. 😅

🕷️

agb32 commented:
ok - lua_toboolean now done, and I think we're probably there.  
I was looking for something like luaL_toboolean, which is why I didn't find it!  
I'll mark the comments as resolved.

agb32 commented:
Right, I think that's all done.  
I've not updated the messages in shadowvigil.lua, because I don't think I'd do that well.  Better for someone who knows more about it to do it... though I've left them there as a placeholder.
Perhaps, once this is pulled, then someone could work on that straight away.

onpon4 commented on src/nlua_misn.c:
```diff
@@ -109,6 +112,8 @@ static const luaL_Reg misn_methods[] = {
    { "npcRm", misn_npcRm },
    { "claim", misn_claim },
+   { "setLogID", misn_setLogID },
+   { "getLogID", misn_getLogID },
```
Are these needed for something? I see that you used them in the Shadow Vigil logs you specified but they seem to be unnecessary, i.e. it seems simple variables would do just as well.

🕵️

agb32 replied:
Yes, required to set/get the logid for the mission, which will ensure that it is saved correctly.

onpon4 replied:
Oh, is it because it can change between saves? Hm... it would be better for that to be transparently handled in the background (as it is for pilots for example). Would there be any way to do that, so that you could just store the log ID or whatever in a variable and rely on it continuing to work through saves? That's how we do things in the rest of the game, requiring this different way is confusing and prone to errors.

Alternatively it might be better to merge "creating" the log and "appending" the log into one function? So that it would just look up the proper log by name at the appropriate time. Basically you would just have appendLog include the identifier string as an argument instead of log ID and it would find out the log ID purely through C code (and not expose that to the Lua scripter).

Anyway, just some thoughts. If no change is feasible this quirk at least needs to be documented.

🕵️

onpon4 commented on src/nlua_shiplog.c:
```diff
@@ -0,0 +1,140 @@
+ *    @luatparam string idstr ID string to identify this log, or empty string for unnamed logsets.
+ *    @luatparam string name Name for this log.
+ *    @luatparam string type Type of log (e.g travel, trade, etc, can be anything.)
+ *    @luatparam boolean overwrite Whether to overwrite existing mission with this logname and logtype.  Warning, removes previous entries of this logname and type if 1, or logs of this type if 2.
```
Some of this explanation is outdated, particularly the thing about a mission or whatever and the mention of "1" and "2".

🕵️

agb32 replied:
Done...

onpon4 commented on src/nlua_shiplog.c:
```diff
@@ -0,0 +1,140 @@
+ *    @luatparam string type Type of log (e.g travel, trade, etc, can be anything.)
+ *    @luatparam boolean overwrite Whether to overwrite existing mission with this logname and logtype.  Warning, removes previous entries of this logname and type if 1, or logs of this type if 2.
+ *    @luatreturn number The logid of the mission.
+ * @luafunc createLog( logname, logtype, overwrite )
```
Args list is out of date (should be `idstr, name, type, overwrite`)

🕵️

agb32 replied:
Done (and added maxLen

onpon4 commented on src/nlua_shiplog.c:
```diff
@@ -0,0 +1,140 @@
+/**
+ * @brief Appends to the shiplog.
+ *
+ *    @luatparam string message Message to append to the log.
```
logid param is missing. That's still needed, right?

🕵️

agb32 replied:
Yes, added.

onpon4 commented on src/nlua_shiplog.c:
```diff
@@ -0,0 +1,140 @@
+ * @file nlua_shiplog.c
+ *
+ * @brief Handles the shiplog Lua bindings.
+ */
```
You could use this comment section to show a full example of log usage in between `@code` and `@endcode` tags. Take a look at the other modules for an example.

🕵️

agb32 replied:
Example added.

onpon4 commented:
Would it be possible to implement that pruning mechanic I mentioned? Not essential but would be useful for the travel logs thing

🕷️

agb32 commented:
I've added the pruning mechanic.   Now with createLog(), you can specify a maximum length for the log, and if it gets longer than this, the earliest entries get removed.

onpon4 commented on src/nlua_shiplog.c:
```diff
@@ -0,0 +1,149 @@
+ *    @luatparam string name Name for this log.
+ *    @luatparam string type Type of log (e.g travel, trade, etc, can be anything.)
+ *    @luatparam boolean overwrite Whether to overwrite existing mission with this logname and logtype.  Warning, removes previous entries of this logname and type.
+ *    @luatparam int maxLen Maximum length of the log (zero for infinite) - if greater than this length, new entries appended will result in old entries being removed.
```
The type specified should be "number" rather than "int" (Lua doesn't differentiate integers from floats), and `nil` should be used for infinite rather than zero, and the argument noted as optional (`nil` is the default of any argument / variable in Lua, so making it infinite with `nil` makes the argument optional and causes a maximum length to be recorded only if specified).

agb32 replied:
ok - I've changed to nil using the code below.

onpon4 commented on src/nlua_shiplog.c:
```diff
@@ -0,0 +1,149 @@
+   logname    = luaL_checkstring(L,2);
+   logtype    = luaL_checkstring(L,3);
+   overwrite = lua_toboolean(L,4);
+   maxLen = luaL_checkint(L,5);
```
I believe (to do what I commented on the doc line), you would do something like:

```
maxLen = 0;
if ( lua_gettop(L) > 4 )
   maxLen = MAX(0, luaL_checkint(L,5));
```

agb32 replied:
Thanks - I've added this.

onpon4 commented:
OK, so just to expand on something I said elsewhere: would there be any particular reason to not do it so it works like this?

`appendLog( idstr, message )`

And then it could use idstr to grab the appropriate ID or whatever internally? That would avoid the potential problems misn.setLogID and misn.getLogID are trying to solve (and make sure that variables are robust).

🕵️

agb32 commented:
Its already implemented with the logid, which will be more efficient (int comparison, rather than string) - (I know this is only a small effect).
And its easier to generate the next in sequence for a logid, than for an idstr, i.e. for missions that don't want an idstr.
So, logid is used to define the logset, while idstr is used to couple different missions together (and thus create or obtain the logid).

onpon4 commented:
More efficient yes, but variables are supposed to be reliable between saves. You said that misn.setLogID and misn.getLogID are essential to prevent problems, yes? That goes against how the Lua is supposed to be handled. Lua scripters should essentially be unaware of the idea of sessions; it should be as if saving, leaving, and then loading the game just brings the Lua code back to how it was. If you need to recapture a log ID, that goes against that.

True that it's more efficient but it's probably not worth the potential for errors and confusion.

🕵️

agb32 commented:
Right - so originally, I had it all transparent in the nlua_misn, but had to remove that and put it in nlua_shiplog, so that it could be used for events.  So, what I need is, in nlua_misn, a misn.createLog() that creates the log and saves the logid.
appendLog could then also be in misn.  And we'd have it separately for events.  But again, same problem, we need to save the logid.

We could do it all with idstr, but that has the potential to create confusion by different missions or events using the same idstr unintentionally.

However, I really want to get this wrapped up now, so if you're sure that forcing a idstr is the way to go, I'll do that.  That way, the event or mission would save nothing specific, and appendLog() would be appendLog(idstr,msg), rather than (logid, msg).

onpon4 commented:
> We could do it all with idstr, but that has the potential to create confusion by different missions or events using the same idstr unintentionally.

I don't think that's a problem. You're honestly overthinking this, we've been using the `var` module with plain strings for a long time with no major problems. If there's a conflict we can always just fix said conflict. It's a lot better than making the system overly complex in my opinion.

> However, I really want to get this wrapped up now, so if you're sure that forcing a idstr is the way to go, I'll do that. That way, the event or mission would save nothing specific, and appendLog() would be appendLog(idstr,msg), rather than (logid, msg).

Yeah, I think that would be ideal.

🕵️

agb32 commented:
ok, I'll do that...

agb32 commented:
ok - done - updated the shadowvigil.lua too.  Its now just:
    shiplog.createLog("Shadow","Shadow vigil","Shadow mission",true);
    shiplog.appendLog("Shadow",misn_desc)

(or for pruning, add a number after the true).

onpon4 commented:
Awesome! Sounds good! 😄

So now the project members just need to get caught up on the pile of pull requests that have built up. 😛

🕷️

agb32 commented:
(fixed an old comment)
